src/protocol/dot_product.rs

```diff
@@ -0,0 +1,173 @@
+///
+/// This extension is summarized in the paper:
+/// "Fast Large-Scale Honest-Majority MPC for Malicious Adversaries"
+/// by Koji Chida, Daniel Genkin, Koki Hamada, Dai Ikarashi, Ryo Kikuchi, Yehuda Lindell, and Ariel Nof
+/// <https://link.springer.com/content/pdf/10.1007/978-3-319-96878-0_2.pdf>
+///
+/// In that paper, this functionality is called `F_product` and
+/// an instantiation for Replciated Secret Sharing is provided in section 6.1 on page 27
+///
+/// To summarize the CHIKP protocol:
+/// each helper computes `u_i+1 = (s_i + s_i+1) * (t_i + t_i+1) - s_i+1 * t_i+1`
+/// It then randomizes this share with randomness that it shares with the
+/// helper to its left, and sends it to the helper to its right,
+/// and receives `u_i` from the helper to its left.
+/// Each helper then defines the pair `(u_i, u_i+1)` as its shares of the output.
+///
+/// (Note it is equivalent to compute `u_i+1 = s_i * t_i + s_i+1 * t_i + s_i * t_i+1`,
+/// but this requires three multiplications instead of two. Assuming multiplying field elements is
+/// significantly more costly (in CPU terms) than addition, this is less efficient to compute)
+///
+/// The key observation that makes this protocol work, is that each helper can compute many `u_i+1` values
+/// from many multiplications, sum them all together, randomize them with correlated randomness, and can
+/// send this single field value to the next helper. The result will be a replicated secret sharing of the
+/// dot-product of the two vectors.
+///
+pub mod stream {
+    use crate::{
+        error::BoxError,
+        field::Field,
+        helpers::{fabric::Network, Direction},
+        protocol::{context::ProtocolContext, RecordId},
+        secret_sharing::Replicated,
+    };
+    use futures::stream::StreamExt; // for `next`
+    use futures::Stream;
+    use std::pin::Pin;
+
+    use serde::{Deserialize, Serialize};
+
+    /// A message sent by each helper when they've computed one share of the result
+    #[derive(Debug, Serialize, Deserialize, Default)]
+    pub struct UValue<F> {
+        u: F,
+    }
+
+    type PairOfShares<F> = (Replicated<F>, Replicated<F>);
+
+    ///
+    /// Consumes the input stream of pairs of replicated secret shares and
+    /// for each pair, multiplies them, and accumulates the result in a running sum.
+    /// once all the values in the stream are consumed, it randomizes the sum with
+    /// randomness it shares with the helper to the left, and sends to the right.
+    /// It receives a randomized sum from the helper to its left, and is now in
+    /// possession of a replicated secret sharing of the dot-product of the two
+    /// streaming vectors of replicated secret shares.
+    ///
+    /// ## Errors
+    /// Lots of things may go wrong here, from timeouts to bad output. They will be signalled
+    /// back via the error response
+    #[allow(dead_code)]
+    pub async fn accumulate_dot_product<F: Field, N: Network>(
```

Comment:
i think there is a way to use stream without `Pin<Box<dyn Stream>>`. roughly speaking, it would look something like:

```rust
pub async fn accumulate_dot_product<F: Field, N: Network, S: Stream<Item = PairOfShares<F>>(
    ctx: ProtocolContext<'_, N>,
    record_id: RecordId,
    input_stream: S,
) -> Result<Replicated<F>, BoxError> {
    ...

    // instead of `while let Some(item) = input_stream.next()`
    // use `fold`: https://docs.rs/futures/latest/futures/stream/trait.StreamExt.html#method.fold
    let u_right = input_stream.fold(s_right - s_left, |acc, (a_share, b_share)| async move { 
        let (a_left, a_right) = a_share.as_tuple();
        let (b_left, b_right) = b_share.as_tuple();
        acc + (a_left + a_right) * (b_left + b_right) - a_right * b_right
    }).await
    ...
```
`Pin` is a pain to work with directly, and `.fold` will hide that complexity away.

Reply:
Oh my gosh, thank you so much! This is a million times better. I am really happy to not have to touch this pin / box stuff =).